core/client.go

```diff
@@ -0,0 +1,835 @@
+
+	ch2 := make(chan types.Log)
+
+	sub, err := c.Client.SubscribeFilterLogs(ctx, q, ch2)
+	if err != nil {
+		c.logger.Error("Failed to subscribe to logs", "err", err)
+		return nil, err
+	}
+	c.logger.Info("Subscribed to logs")
+
+	safeSub := NewSafeSubscription(sub)
+	lastBlock := currentBlock
+
+	resubFilterLogs := func() error {
+		currentBlock, err := c.Client.BlockNumber(ctx)
+		if err != nil {
+			c.logger.Error("Failed to get current block number", "err", err)
+			return err
+		}
+		c.logger.Debug("Got current block number for resub", "block", currentBlock)
+
+		fromBlock := max(lastBlock+1, currentBlock-BLOCK_MAX_RANGE)
+
+		for fromBlock < currentBlock {
+			if fromBlock > currentBlock {
+				break
+			}
+
+			toBlock := fromBlock + BLOCK_CHUNK_SIZE
+
+			targetBlock := big.NewInt(int64(toBlock))
+			if fromBlock+BLOCK_CHUNK_SIZE > currentBlock {
+				targetBlock = nil
+			}
+
+			logs, err := c.Client.FilterLogs(ctx, ethereum.FilterQuery{
+				FromBlock: big.NewInt(int64(fromBlock)),
+				ToBlock:   targetBlock,
+				Addresses: q.Addresses,
+				Topics:    q.Topics,
+			})
+			if err != nil {
+				c.logger.Error("Failed to get missed logs", "err", err)
+				return err
+			} else {
+				c.logger.Info("Got missed logs on resubscribe", "count", len(logs))
+				for _, log := range logs {
+					ch2 <- log
+				}
+			}
+
+			fromBlock += BLOCK_CHUNK_SIZE
+		}
+
+		return nil
+	}
+
+	resub := func() error {
+		c.clientLock.RLock()
+		defer c.clientLock.RUnlock()
+
+		err := resubFilterLogs()
+		if err != nil {
+			return err
+		}
+
+		sub, err = c.Client.SubscribeFilterLogs(ctx, q, ch2)
+		if err != nil {
+			c.logger.Error("Failed to resubscribe to logs", "err", err)
+			return err
+		}
+		c.logger.Info("Resubscribed to logs")
+
+		safeSub.SetUnderlyingSub(sub)
+
+		return nil
+	}
+
+	lastBlock = currentBlock
+
+	c.wg.Add(1)
+	go func() {
+		defer c.wg.Done()
+
+		ticker := time.NewTicker(RESUB_INTERVAL)
+		defer ticker.Stop()
+
+		reinitC := c.WatchReinit()
+
+		for {
+			select {
+			case <-safeSub.Err():
+				c.logger.Debug("Safe subscription ended")
+				return
+			case success := <-reinitC:
+				reinitC = c.WatchReinit()
+				if success {
+					err := resub()
+					c.handleClientError(err)
+				}
+			case log := <-ch2:
+				lastBlock = max(lastBlock, log.BlockNumber)
+				ch <- log
+			case <-ticker.C:
+				c.logger.Debug("Resub ticker fired")
+				err := resub()
+				c.handleClientError(err)
+			case <-sub.Err():
+				c.logger.Info("Underlying subscription ended, resubscribing")
+				err := resub()
+				c.handleClientError(err)
+			case <-ctx.Done():
+				c.logger.Info("Context done, ending subscription")
+				safeSub.Unsubscribe()
+				return
+			}
+		}
+	}()
+
+	return safeSub, nil
+}
+
+func (c *SafeEthClient) Close() {
+	c.clientLock.Lock()
+	defer c.clientLock.Unlock()
+
+	close(c.closeC)
+	c.wg.Wait()
+	c.logger.Info("SafeEthClient closed")
+}
+
+func (c *SafeEthClient) isConnectionError(err error) bool {
```

Comment:
Honestly, since this is the only utility and it's private, I prefer to leave it in the same file and use a receiver for consistency anyway. If we have more than one utility there, we can just move them to a `core/safeclient/utils.go` and make them non-receivers.

Reply:
We can still leave it in the same file. But receiver is pointless here. It also attaches this method to a particular object while it doesn't need it, or we have to call on nil.

> If we have more than one utility
It's much harder to notice if they have receiver, it's just confusing. 

Tried to find something in style guide on this found only [this](https://www.reddit.com/r/golang/comments/112qvko/to_receive_or_not_to_receive/) thread on reddit

Not sure what do you mean by consistency otherwise all utils will have to follow the same pattern.

Reply:
As discussed today, agree with the understanding that it's better to use receivers whenever it's actually a receiver, my issue with it was that:
1. Having a one-function `utils.go` would be a bit excessive
2. Just throwing a standalone function between lots of receivers is confusing for whoever is reading that

That said, now we did have some more logic that could be more acceptably moved to utilities, so done in [c9569bc](https://github.com/NethermindEth/near-sffl/pull/154/commits/c9569bc1c565aa8792fc25dd903f999f5aa4eef2)